From 2f49fba8fff6c2af424aeb8c752c4bd718893220 Mon Sep 17 00:00:00 2001
From: Pranoy Goru <gorumanip@ami.com>
Date: Fri, 14 Jul 2023 10:22:25 +0530
Subject: [PATCH 1/2] Restricted root user privilage

Description :

Restricted root user privilage for RoleId and User delete method

Test Cases:

1) patch https://{{ip}}/redfish/v1/AccountService/Accounts/root

UserName : root
Password : XXXX

Body Json

{
    "RoleId": "Administrator"
}

response 403Forbidden

{
    "error": {
        "@Message.ExtendedInfo": [
            {
                "@odata.type": "#Message.v1_1_1.Message",
                "Message": "While attempting to establish a connection to 'redfish/v1/AccountService/Accounts/root', the service denied access.",
                "MessageArgs": [
                    "redfish/v1/AccountService/Accounts/root"
                ],
                "MessageId": "Base.1.13.0.AccessDenied",
                "MessageSeverity": "Critical",
                "Resolution": "Attempt to ensure that the URI is correct and that the service has the appropriate credentials."
            }
        ],
        "code": "Base.1.13.0.AccessDenied",
        "message": "While attempting to establish a connection to 'redfish/v1/AccountService/Accounts/root', the service denied access."
    }
}

2) Delete https://{{ip}}/redfish/v1/AccountService/Accounts/root

UserName : root
Password : XXXX

response 405Method Not Allowed

{
    "error": {
        "@Message.ExtendedInfo": [
            {
                "@odata.type": "#Message.v1_1_1.Message",
                "Message": "The delete request failed because the resource requested cannot be deleted.",
                "MessageArgs": [],
                "MessageId": "Base.1.13.0.ResourceCannotBeDeleted",
                "MessageSeverity": "Critical",
                "Resolution": "Do not attempt to delete a non-deletable resource."
            }
        ],
        "code": "Base.1.13.0.ResourceCannotBeDeleted",
        "message": "The delete request failed because the resource requested cannot be deleted."
    }
}

Signed-off-by: Pranoy Goru <gorumanip@ami.com>
---
 redfish-core/lib/account_service.hpp | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/redfish-core/lib/account_service.hpp b/redfish-core/lib/account_service.hpp
index 5ebe8695..29ac15e6 100644
--- a/redfish-core/lib/account_service.hpp
+++ b/redfish-core/lib/account_service.hpp
@@ -1219,7 +1219,16 @@ inline void updateUserProperties(std::shared_ptr<bmcweb::AsyncResp> asyncResp,
                     dbus::utility::DbusVariantType{*enabled});
             }
 
-            if (roleId)
+            if((username == "root") && roleId)
+            {
+                BMCWEB_LOG_ERROR << "Not able to change privilage level for root user";
+		const std::string& arg = "redfish/v1/AccountService/Accounts/" + username;
+                messages::accessDenied(
+                                  asyncResp->res,
+                                  boost::urls::format(arg));
+                return;
+            }
+	    else if (roleId)
             {
                 std::string priv = getPrivilegeFromRoleId(*roleId);
                 if (priv.empty())
@@ -1963,6 +1972,13 @@ inline void
     tempObjPath /= username;
     const std::string userPath(tempObjPath);
 
+    if (username == "root")
+    {
+	BMCWEB_LOG_DEBUG << "Not able to delete root user\n" ;
+	messages::resourceCannotBeDeleted(asyncResp->res);
+	return;
+    }
+
     crow::connections::systemBus->async_method_call(
         [asyncResp, username](const boost::system::error_code& ec) {
         if (ec)
-- 
2.25.1

